### PR TITLE
feat(sme): implement websocket entrypoint

### PR DESCRIPTION
**Pull request by jhaeyaert**
## Issue

https://github.com/gravitee-io/issues/issues/8253

## Description

Implements the WebSocket entrypoint connector for SME.

This required to rework the previous support of WebSocket with Jupiter engine (v3 compatibility backport, see https://github.com/gravitee-io/gravitee-api-management/pull/2151) in order to introduce a dedicated `WebSocket` interface that supports reactive and add a `WebSocketAdapter`.

As a consequence, reviewing this PR requires:

- Validate the websocket behavior with Jupiter engine and Jupiter v3 compatibility mode.
- Validate the WebSocket entrypoint that work with SME for async api (message oriented).

The following PR must be included in the review process:

- gateway-api: add the new WebSocket interface (https://github.com/gravitee-io/gravitee-gateway-api/pull/147)
- http-connector: fix v3 `connected()` method (https://github.com/gravitee-io/gravitee-connector-http/pull/40)

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/8253-websocket-entrypoint/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-zknkqnjfat.chromatic.com)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-http/src/test/java/io/gravitee/gateway/jupiter/http/vertx/VertxHttpServerRequestTest.java [hunk 1]
```diff
@@ -372,19 +372,19 @@ void shouldCreateWebSocket() {
...
-        when(httpServerRequest.method()).thenReturn(HttpMethod.GET);
-        when(httpServerRequest.version()).thenReturn(HttpVersion.HTTP_1_1);
-
-        final WebSocket webSocket = cut.webSocket();
-
-        for (int i = 0; i < 10; i++) {
-            assertSame(webSocket, cut.webSocket());
-        }
-    }
+    //    @Test
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-http/src/test/java/io/gravitee/gateway/jupiter/http/vertx/VertxHttpServerRequestTest.java]: Should we re-enable those tests?

File: gravitee-apim-gateway/gravitee-apim-gateway-http/src/test/java/io/gravitee/gateway/jupiter/http/vertx/VertxHttpServerRequestTest.java [hunk 2]
```diff
@@ -407,33 +407,33 @@ void shouldNotBeWebSocketUpgradedWhenWebSocketNotCreated() {
...
-        when(httpServerRequest.method()).thenReturn(HttpMethod.GET);
-        when(httpServerRequest.version()).thenReturn(HttpVersion.HTTP_1_1);
-
-        final WebSocket webSocket = cut.webSocket();
-        ReflectionTestUtils.setField(webSocket, "upgraded", true);
-
-        assertNotNull(webSocket);
-        assertTrue(cut.isWebSocketUpgraded());
-    }
+    //    @Test
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-http/src/test/java/io/gravitee/gateway/jupiter/http/vertx/VertxHttpServerRequestTest.java]: Same here ?

File: gravitee-apim-plugin/gravitee-apim-plugin-entrypoint/gravitee-apim-plugin-entrypoint-websocket/src/main/java/io/gravitee/plugin/entrypoint/websocket/WebSocketEntrypointConnector.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    protected final WebSocketEntrypointConnectorConfiguration configuration;
+
+    public WebSocketEntrypointConnector(final WebSocketEntrypointConnectorConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    @Override
+    public ApiType supportedApi() {
+        return SUPPORTED_API;
```
Comment (guillaumelamirand) [inline gravitee-apim-plugin/gravitee-apim-plugin-entrypoint/gravitee-apim-plugin-entrypoint-websocket/src/main/java/io/gravitee/plugin/entrypoint/websocket/WebSocketEntrypointConnector.java]: Not required as EntrypointAsyncConnector already implemented it.

File: gravitee-apim-plugin/gravitee-apim-plugin-entrypoint/gravitee-apim-plugin-entrypoint-websocket/src/main/resources/schema/schema-form.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "object",
+  "id": "urn:jsonschema:io:gravitee:plugin:entrypoint:webhook:configuration:WebsocketEntrypointConnectorConfiguration",
+  "properties": {},
+  "required": [
```
Comment (guillaumelamirand) [inline gravitee-apim-plugin/gravitee-apim-plugin-entrypoint/gravitee-apim-plugin-entrypoint-websocket/src/main/resources/schema/schema-form.json:32]: You miss the new subscriber/publisher settings you have added.

**Review by guillaumelamirand — CHANGES_REQUESTED** [review]

File: gravitee-apim-gateway/gravitee-apim-gateway-http/src/test/java/io/gravitee/gateway/jupiter/http/vertx/VertxHttpServerRequestTest.java (hunk 1 above)
Comment (jhaeyaert) [reply gravitee-apim-gateway/gravitee-apim-gateway-http/src/test/java/io/gravitee/gateway/jupiter/http/vertx/VertxHttpServerRequestTest.java]: I was just annoyed by intellij while testing another stuffs and didn't take time to uncomment and fix it 🤦🏻‍♂️
Good catch 👍

File: gravitee-apim-plugin/gravitee-apim-plugin-entrypoint/gravitee-apim-plugin-entrypoint-websocket/src/main/java/io/gravitee/plugin/entrypoint/websocket/WebSocketEntrypointConnectorFactory.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ */
+@Slf4j
+@AllArgsConstructor
+public class WebSocketEntrypointConnectorFactory implements EntrypointAsyncConnectorFactory {
+
+    private final ConnectorFactoryHelper connectorFactoryHelper;
+
+    @Override
+    public ApiType supportedApi() {
+        return ApiType.ASYNC;
```
Comment (guillaumelamirand) [inline gravitee-apim-plugin/gravitee-apim-plugin-entrypoint/gravitee-apim-plugin-entrypoint-websocket/src/main/java/io/gravitee/plugin/entrypoint/websocket/WebSocketEntrypointConnectorFactory.java]: Same here, EntrypointAsyncConnectorFactory already defined the api type.

**Review by guillaumelamirand — APPROVED** [review]

**Pull request merged**